### DISCUSSION TITLE
unregistered namespace

### DISCUSSION THREAD

**Discussion by mlandsfeld** [tethysplatform/tethys#706] [Q&A]
I am installing my app on a new portal and when I navigate to the Apps page, in the portal, I get the following error:

NoReverseMatch at /apps/

'geoserver_app' is not a registered namespace
Request Method:	GET
Request URL:	http://chc-tethys.chc.ucsb.edu/apps/
Django Version:	2.2.12
Exception Type:	NoReverseMatch
Exception Value:	
'geoserver_app' is not a registered namespace
Exception Location:	/home/tethys/anaconda3/envs/tethys/lib/python3.7/site-packages/django/urls/base.py in reverse, line 86
Python Executable:	/home/tethys/anaconda3/envs/tethys/bin/python
Python Version:	3.7.6
Python Path:	
['/home/tethys/anaconda3/envs/tethys/lib/python3.7/site-packages/tethys_portal',
 '/home/tethys/anaconda3/envs/tethys/lib/python37.zip',
 '/home/tethys/anaconda3/envs/tethys/lib/python3.7',
 '/home/tethys/anaconda3/envs/tethys/lib/python3.7/lib-dynload',
 '/home/tethys/anaconda3/envs/tethys/lib/python3.7/site-packages']

Any ideas on what might be causing this? I have two apps and I get the same results after I install either one of them. I had one of the apps installed on another portal and it started up without any issues.
Thanks!

**Comment by msouff** [comment]
Are you installing the apps using the "tethys install" command?

If you have a repo you can share for those apps that would help.

**Comment by ezrajrice** [comment]
I can confirm that trying to install using the "python setup.py install"
command results in this error. Michael's answer above is the correct way of
installing apps now.

Ezra Rice

**Reply by mlandsfeld** [reply]
I am using 'tethys install' from the apps base directory and that seems to work fine. I don't have it on a repo yet but I have it zipped up here:
https://data.chc.ucsb.edu/people/marty/tethysapp-geoserver_app.zip

**Reply by msouff** [reply]
What version of tethys do you have?

I installed your app in my local version and it displays fine

![image](https://user-images.githubusercontent.com/17328135/125496791-2c38aa9e-62f2-41e5-81d0-8e19bdb1b8f1.png)

Since you seem to be in a production environment. Make sure you are running "tethys manage collectall" after you install the app and that you are restarting the server "sudo supervisorctl restart all"

Iif that doesn't work, I would try making sure that the app is uninstalled by running "tethys uninstall <app_name>" and then make sure the app package doesn't show when you run conda list. If it does show, run pip uninstall "<app_name>". This is just a guess but there may be remnants of a previous installation preventing you  from reinstalling the app properly.

**Reply by mlandsfeld** [reply]
I am running Tethys 3.0.5 with Python 3.7.6. I just spent and hour and a half with Billy Ashmall uninstalling and reinstalling and collecting all and we can not get it to run. It ran on two of his installations as well. The only difference from your instructions is that I was running 'sudo systemctl restart nginx' since running superviorctl returns a 'unix:///tmp/supervisor.sock no such file' error. We were thinking that I may need to reinstall tethys as it appears there is something wrong with my installation. Would you recommend the same?
Thanks!

**Reply by msouff** [reply]
Before you reinstall Tethys, I'd try creating a template app with the "tethys scaffold" command. If that template app gives you the same error then that's a good indication that the problem is coming from your tethys installation, not the app.

**Reply by mlandsfeld** [reply]
I was able to install a test_app with tethys scaffold and run the app on my server. I did not get the registered namespace error when going to the app page in the portal. But when I tried to open the geoserver app, I get this printout:

-------------

TemplateDoesNotExist at /apps/geoserver-app/
tethys_apps/app_header_content.html

TemplateDoesNotExist. tethys_apps/app_header_content.html

Django tried loading these templates, in this order:

Using engine django:

django.template.loaders.filesystem.Loader: /home/tethys/anaconda3/envs/tethys/lib/python3.7/site-packages/tethys_portal/templates/tethys_apps/app_header_content.html (Source does not exist)

---------------

That is because there is no tethys_apps directory in  /home/tethys/anaconda3/envs/tethys/lib/python3.7/site-packages/tethys_portal/templates/. Was the installation supposed to make that header file?

**Reply by swainn** [reply]
The additional base templates for apps were not added until Tethys 3.1. See: http://docs.tethysplatform.org/en/stable/whats_new/prior_releases.html#additional-base-templates-for-apps

**Comment by swainn** [comment]
@mlandsfeld this issue can often be caused by an import error in a controller or other file in the app. Another thing to try before reinstalling would be to try importing various modules of your app in a Django shell session. To start the Django shell session run:

```bash
conda activate tethys
python /path/to/miniconda/envs/tethys/lib/pythonX.Y/site-packages/tethys_portal/manage.py shell
from tethysapp import geoserver_app
from tethysapp.geoserver_app import app
from tethysapp.geoserver_app import controllers
```

A few things to check:
* Are all dependencies installed? (conda list)
* Is there another app named geoserver_app installed (that's the name of one of the Tethys tutorial apps I think).
* Does it work if you use `tethys install -d` instead of `tethys install`?

**Reply by mlandsfeld** [reply]
I was able to start a Django shell and install all of those packages. I also uninstalled and reinstalled the geoserver_app with tethys install -d.  But I have similar results as without the -d but a different does not exist:

/home/tethys/anaconda3/envs/tethys/lib/python3.7/site-packages/tethys_portal/templates/geoserver_app/home.html

There is no geoserver_app directory in templates.

**Reply by swainn** [reply]
@mlandsfeld are you no longer getting the `'geoserver_app' is not a registered namespace` error? Sounds like the templates are not being copied over when the files are installed. Can you share with me your `setup.py`?

**Reply by swainn** [reply]
The Google Earth Engine tutorial includes a section that describes the process of preparing and publishing an app for production deployment. Reading through these sections may help:

- [Part 3 Primer: Publishing and Deploying](http://docs.tethysplatform.org/en/stable/tutorials/google_earth_engine/part_3/primer.html)
- [7. Copy Resource Files](http://docs.tethysplatform.org/en/stable/tutorials/google_earth_engine/part_3/prepare.html#copy-resource-files)

We also have an app installation guide here:
- [App Installation Guide](http://docs.tethysplatform.org/en/stable/installation/production/app_installation.html#installing-apps-production)

**Reply by mlandsfeld** [reply]
Hi Nathan, Here is my setup.py code:

from setuptools import setup, find_namespace_packages
from tethys_apps.app_installation import find_resource_files

# -- Apps Definition -- #
app_package = 'geoserver_app'
release_package = 'tethysapp-' + app_package

# -- Python Dependencies -- #
dependencies = []

# -- Get Resource File -- #
resource_files = find_resource_files('tethysapp/' + app_package + '/templates', 'tethysapp/' + app_package)
resource_files += find_resource_files('tethysapp/' + app_package + '/public', 'tethysapp/' + app_package)
resource_files += find_resource_files('tethysapp/' + app_package + '/workspaces', 'tethysapp/' + app_package)


setup(
    name=release_package,
    version='0.0.1',
    description='A test geoserver app',
    long_description='',
    keywords='',
    author='Martin Landsfeld',
    author_email='',
    url='',
    license='',
    packages=find_namespace_packages(),
    package_data={'': resource_files},
    include_package_data=True,
    zip_safe=False,
    install_requires=dependencies,
)

I have followed the instructions for preparing and installing the tethys apps. Thanks for the help!

**Reply by swainn** [reply]
@mlandsfeld your `setup.py` looks ok. Here's a few more questions:

1. What is the result of `tethys list`?
2. Is there a `__init__.py` in the `tethysapp` directory? If so, it needs to be removed.

**Reply by mlandsfeld** [reply]
I did have a __init__.py file in my tethysapp/geoserver_app folder. So I need to uninstall and reinstall now?

Here are the results of tethys list before removing __init__.py:

$ tethys list
Loading Tethys Extensions...
Loading Tethys Apps...
Tethys Apps Loaded: test_app, geoserver_app

Apps:
  test_app
  geoserver_app


And after:

$ tethys list
Loading Tethys Extensions...
Loading Tethys Apps...
Tethys Apps Loaded: test_app

Apps:
  test_app

**Reply by swainn** [reply]
The `__init__.py` file in `tethysapp/geoserver_app` needs to be there. But there should be no `__init__.py` file in the `tethysapp` folder. That folder should only contain the `geoserver_app` folder.
